File: backend/app/schemas/order.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.core.enums import OrderStatus


class OrderItemIn(BaseModel):
    menu_item_id: uuid.UUID
    quantity: int = Field(gt=0, le=50)


class PlaceOrderRequest(BaseModel):
    restaurant_id: uuid.UUID
    items: list[OrderItemIn] = Field(min_length=1)
    delivery_note: str | None = Field(default=None, max_length=500)
    tip_minor: int = Field(default=0, ge=0)


class OrderItemOut(BaseModel):
    id: uuid.UUID
    menu_item_id: uuid.UUID
    name_snapshot: str
    unit_price_minor: int
    quantity: int


class OrderEventOut(BaseModel):
    id: uuid.UUID
    status: OrderStatus
    occurred_at: datetime
    note: str | None


class OrderOut(BaseModel):
    id: uuid.UUID
    client_local_id: str
    restaurant_id: uuid.UUID
    status: OrderStatus
    total_minor: int
    currency: str
    eta: datetime | None
    placed_at: datetime
    delivery_note: str | None
    tip_minor: int
    route_polyline: str | None
    version: int
    updated_at: datetime
    deleted: bool = False
    items: list[OrderItemOut] = []
    events: list[OrderEventOut] = []


class CancelOrderRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=500)

File: android/core/model/src/main/kotlin/com/ordertracking/core/model/Order.kt
package com.ordertracking.core.model

import java.time.Instant

/**
 * Pure domain model -- no Room/Android annotations here (those live on
 * OrderEntity in :core:database). [localId] is the identity that matters:
 * it's a client-generated UUID that exists before the server has ever heard
 * of this order, which is what makes offline creation a non-special case
 * (DESIGN.md §4).
 */
data class Order(
    val localId: String,
    val serverId: String?,
    val restaurantId: String,
    val status: OrderStatus,
    val serverVersion: Long,
    val placedAtLocal: Instant,
    val serverUpdatedAt: Instant?,
    val totalMinor: Long,
    val currency: String,
    val syncState: SyncState,
    val lastError: String?,
    val etaAtServer: Instant?,
    val deliveryNote: String?,
    val tipMinor: Long,
    val routePolyline: String?,
    val items: List<OrderItem>,
    val events: List<OrderEvent>,
)

data class OrderItem(
    val id: String,
    val orderLocalId: String,
    val menuItemId: String,
    val nameSnapshot: String,
    val unitPriceMinor: Long,
    val quantity: Int,
)

data class OrderEvent(
    val id: String,
    val orderLocalId: String,
    val status: OrderStatus,
    val occurredAt: Instant,
    val note: String?,
)
